parlai/agents/language_model/language_model.py

```diff
@@ -0,0 +1,480 @@
+        self.hidden = self.repackage_hidden(hidden)
+        # get only last predicted word
+        output = output.select(0,-1)
+        output_flat = output.view(-1, len(self.dict))
+        loss += y_lens_batch[0]*self.criterion(output_flat, targets.select(1,0).view(-1)).data
+
+        for i in range(1, targets.size(1)):
+            output, hidden = self.model(targets.select(1,i-1).view(1, bsz), self.hidden, no_pack=True)
+            self.hidden = self.repackage_hidden(hidden)
+            output_flat = output.view(-1, len(self.dict))
+            loss += y_lens_batch[i]*self.criterion(output_flat, targets.select(1,i).view(-1)).data
+
+        return loss/float(sum(y_lens_batch))
+
+    def get_predictions(self, data, end_idxs):
+        """Generates predictions word by word until we either reach the end token
+           or some max length (opt['truncate_pred']).
+        """
+        token_list = []
+        bsz = data.size(0)
+        done = [False for _ in range(bsz)]
+        total_done = 0
+        hidden = self.model.init_hidden(bsz)
+        output, hidden = self.model(data.transpose(0,1), hidden)
+        hidden = self.repackage_hidden(hidden)
+        word_weights = output.squeeze().data.exp()
+        # get last word of output
+        if bsz > 1:
+            value, word_idx = torch.max(word_weights, 2)
+            last_words = []
+            for j in range(len(end_idxs)):
+                last_words.append(word_idx[end_idxs[j]-1][j])
+            word_idx = torch.LongTensor(last_words)
+            if self.use_cuda:
+                word_idx = word_idx.cuda()
+        else:
+            word_weights = word_weights[-1]
+            value, word_idx = torch.max(word_weights, 0)
+        # mark end indices in batch
+        for k in range(word_idx.size(0)):
+            if not done[k]:
+                if int(word_idx[k]) == self.END_IDX:
+                    done[k] = True
+                    total_done +=1
+        token_list.append(word_idx.view(bsz, 1))
+
+        i = 1
+
+        while total_done < bsz and i <= self.opt['truncate_pred']:
+            new_col = word_idx.clone()
+            new_col = Variable(new_col)
+            if self.use_cuda:
+                new_col = new_col.cuda()
+            output, hidden = self.model(new_col.view(1, bsz), hidden, no_pack=True)
+            hidden = self.repackage_hidden(hidden)
+            word_weights = output.squeeze().data.exp()
+            if bsz > 1:
+                value, word_idx = torch.max(word_weights, 1)
+            else:
+                value, word_idx = torch.max(word_weights, 0)
+            # mark end indices for items in batch
+            for k in range(word_idx.size(0)):
+                if not done[k]:
+                    if int(word_idx[k]) == self.END_IDX:
+                        done[k] = True
+                        total_done += 1
+            token_list.append(word_idx.view(bsz, 1))
+            i += 1
```

Comment:
isn't this essentially the same as above? can't we just start with `i = 0` and cut out the previous section?

Reply:
Not exactly. i = 0 is a special case, because we get all of the outputs from the input sequence and have to select the correct tokens from the end indices. I could write this as one big loop and have special cases for i=0, but i thought this was slightly more clear.